src/DotLiquid/Template.cs

```diff
@@ -174,7 +174,7 @@ public static void RegisterSafeType(Type type, string[] allowedMembers, Func<obj
         /// <param name="func">Function that converts the specified type into a Liquid Drop-compatible object (eg, implements ILiquidizable)</param>
         public static void RegisterSafeType(Type type, Func<object, object> func)
         {
-            SafeTypeTransformers[type] = func;
+            SafeTypeTransformers.AddOrUpdate(type, func, (k, oldFunction) => func);

```

Comment:
AddOrUpdate resolves concurrency issues by ensuring atomicity in the add or update operations. It guarantees that:

- If multiple threads attempt to add a new entry for the same key, only one will succeed, and the others will move to the update logic.
-  If multiple threads attempt to update an existing entry, AddOrUpdate ensures that each update is applied on the latest value, one after the other, 
preventing any thread from overwriting changes made by another thread without first seeing those changes.

Reply:
And you are not taking benefit from the update logic because your provided delegate always overwrites the old function without actually taking the old function into consideration.

Reply:
We are still back to square 1: demonstrate the issue through a unit test, leverage that unit test to prove that your change resolves the race condition through an orderly conflict (update) resolution.